src/arch/xtos-wrapper/include/kernel.h

```diff
@@ -0,0 +1,31 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2022 Intel Corporation. All rights reserved.
+ *
+ * Author: Jyri Sarha <jyri.sarha@intel.com>
+ */
+
+#ifndef __XTOS_WRAPPER_KERNEL_H__
+#define __XTOS_WRAPPER_KERNEL_H__
+
+#include <sof/platform.h>
+#include <sof/trace/trace.h>
+#include <user/trace.h>
+
+#include <stddef.h>
+#include <stdint.h>
+
+extern struct tr_ctx wait_tr;
+
+static inline void k_yield(void)
+{
+	tr_dbg(&wait_tr, "WFE");
+#if CONFIG_DEBUG_LOCKS
+	if (lock_dbg_atomic)
+		tr_err_atomic(&wait_tr, "atm");
+#endif
+	platform_wait_for_interrupt(0);
```

Comment:
we're trying to recreate the `k_yield()` from Zephyr, right? The description in Zephyr says:
```
/**
 * @brief Yield the current thread.
 *
 * This routine causes the current thread to yield execution to another
 * thread of the same or higher priority. If there are no other ready threads
 * of the same or higher priority, the routine returns immediately.
 */
```
Are you sure WFI is what we need? In Zephyr `k_yield()` will **return immediately** whereas a WFI will suspend the thread until an interrupt. Don't we need something like a call to the scheduler instead?

Reply:
Thanks @jsarha it looks like you have found that all wait_for_interrupt() users were xtos based i.e.
```
src/arch/xtensa/schedule/task.c:38:		wait_for_interrupt(0);
src/arch/xtensa/lib/cpu.c:209:		arch_wait_for_interrupt(0);
src/drivers/intel/cavs/sue-ipc.c:58:			wait_for_interrupt(0);
src/drivers/intel/haswell/ipc.c:111:			wait_for_interrupt(0);
src/drivers/amd/renoir/ipc.c:165:			wait_for_interrupt(0);
src/drivers/mediatek/mt8186/ipc.c:89:		wait_for_interrupt(0);
src/drivers/imx/ipc.c:122:			wait_for_interrupt(0);
src/init/init.c:129:		wait_for_interrupt(0);
src/schedule/task.c:57:		wait_for_interrupt(0);
src/platform/intel/cavs/lps_wait.c:101:		arch_wait_for_interrupt(0);
```
Can you confirm that
```
static inline void wait_delay(uint64_t number_of_clks) {}
static inline void wait_delay_ms(uint64_t ms) {}
static inline void wait_delay_us(uint64_t us) {}
```
all have only xtos or driver users (an no audio/pipeline code). If so we can close this and mark it NA.

Reply:
There appears to be two instances of wait_delay usage in  src/audio/pipeline/pipeline-stream.c:431 and in src/ipc/ipc4/handler.c:375. The rest of usage is in either under drivers or platform subdirs, e.g. only for implementation XTOS. BTW, the ipc4/handler.c usage has a curious comment on it:
`	while (msg_data.delayed_reply) {
		/* TODO: update to yield(tmeout) for Zephyr */
		wait_delay(10000);

		if (!try_count--) {
			ret = IPC4_FAILURE;
			tr_err(&ipc_tr, "failed to wait schedule thread");
			break;
		}
	}
`
@lgirdwood I did not notice this second usage while we talked,  but reading the comment should we still keep yield() wrapper around to for this (then again yeald() api does not take timeout as a parameter)?

Reply:
@jsarha looks like https://docs.zephyrproject.org/3.0.0/reference/kernel/threads/index.html#c.k_sleep will fix both.